benches/hashmap_benchmark.rs

```diff
@@ -0,0 +1,228 @@
+// The benchmarks aim to only measure times of the operations in their names.
+// That's why all use Bencher::iter_batched which enables non-benchmarked
+// preparation before running the measured function.
+// Insert (which doesn't completely avoid updates, but makes them unlikely),
+// remove and search have benchmarks with empty values and with custom structs
+// of 42 64-bit integers.
+// (as a sidenote, the performance really differs; in the case of remove, the
+// remove function itself returns original value - the benchmark doesn't use
+// this value, but performance is significantly worse - about twice on my
+// machine - than the empty value remove; it might be interesting to see if a
+// remove function returning void would perform better, ie. if the returns
+// are optimized - omitted in this case).
+// The counts of inserted/removed/searched elements are chosen at random from
+// constant ranges in an attempt to avoid a single count performing better
+// because of specific HW features of computers the code is benchmarked with.
+
+extern crate concread;
+extern crate criterion;
+extern crate rand;
+
+use concread::hashmap::*;
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
+use rand::{thread_rng, Rng};
+use std::mem;
+
+// ranges of counts for different benchmarks:
+const INSERT_COUNT: (usize, usize) = (120, 140);
+const INSERT_COUNT_FOR_REMOVE: (usize, usize) = (340, 360);
+const REMOVE_COUNT: (usize, usize) = (120, 140);
+const INSERT_COUNT_FOR_SEARCH: (usize, usize) = (120, 140);
+const SEARCH_COUNT: (usize, usize) = (120, 140);
+// In the search benches, we randomly search for elements of a range of SEARCH_SIZE.0/SEARCH_SIZE.1
+// times the number of elements contained.
+const SEARCH_SIZE: (usize, usize) = (4, 3);
+
+pub fn insert_empty_value(c: &mut Criterion) {
+    c.bench_function("insert_empty_value", |b| {
+        b.iter_batched(
+            || {
+                let mut rng = thread_rng();
+                let count = rng.gen_range(INSERT_COUNT.0, INSERT_COUNT.1);
+                let mut list = Vec::with_capacity(count);
+                for _ in 0..count {
+                    list.push(rng.gen_range(0, INSERT_COUNT.1 << 8) as u32)
```

Comment:
You could consider generating a "random set" as a const or a global, and then using that. That way it's a "random access pattern" to the map, but it's "consistent" to us the benchmarkers. You could also have a couple of these random sets too. That way we can see consistent results but still stress random paths.

Reply:
Or using a PRNG with a fixed seed so we always generate the same data stream.

Reply:
Are you sure we want that? I thought it was actually better to use multiple random paths so that occasional deviations due to HW are lost.
Isn't knowing the pattern better for actually noticing certain deviations?

Reply:
i'm not sure if your comment is "for" a seeded prng or against it? 

Thinking about this more I think we can leave it as true rng though, since criterion does re-run tests multiple times to eliminate noise anyway ...